### PR TITLE
Nav category structure

### PR DESCRIPTION
**Pull request by vapaj**
Reads the navbar category structure from lettera. This only adds categories for the "main" nav visible under the HBL logo. The structure under `/meny` is done separately.

### CODE REVIEW

File: apps/mosaico/src/Main.purs [hunk 1]
```diff
@@ -110,31 +120,48 @@ spec ::
...
-      guards = { credentials: getCredentials }
-  Aff.launchAff_ $ Payload.startGuarded (Payload.defaultOpts { port = serverPort }) spec { handlers, guards }
+  Aff.launchAff_ do
+    htmlTemplate <- liftEffect $ FS.readTextFile UTF8 indexHtmlFileLocation
+    categoryStructure <- Lettera.getCategoryStructure HBL
+    categoryRegex <-
+      -- This is used for matching a category label from a route, such as "/nyheter" or "/norden-och-världen"
+      case Regex.regex "^\\/([\\w|ä|ö|å|-]+)\\b" Regex.ignoreCase of
+        Right r   -> pure r
+        Left _err -> Aff.throwError $ Aff.error "Got weird regex, couldn't parse it. Server is now exploding. Please fix it."
```
Comment (kaol) [inline apps/mosaico/src/Main.purs:150]: Or just use some `unsafe` foo to coerce it, having static regex definition string explode is just not likely. Simpler code to not even pattern match.

File: apps/mosaico/src/Mosaico.purs [hunk 2]
```diff
@@ -96,22 +103,35 @@ mosaicoComponent initialValues props = React.do
...
+
   useEffect state.route do
     case state.route of
       Routes.Frontpage -> do
         if null state.frontpageArticles
         then Aff.launchAff_ do
-          frontpage <- Lettera.getFrontpage HBL
-          liftEffect $ setState \s -> s { frontpageArticles = frontpage, article = Nothing }
+          frontpage <- Lettera.getFrontpage HBL Nothing
+          liftEffect $ setState \s -> s { frontpageArticles = frontpage
```
Comment (kaol) [inline apps/mosaico/src/Mosaico.purs:126]: `setState _`

File: apps/mosaico/src/Mosaico.purs [hunk 3]
```diff
@@ -124,6 +144,18 @@ mosaicoComponent initialValues props = React.do
...
       Routes.NotFoundPage _path -> pure unit
+      Routes.CategoryPage category -> do
+        -- TODO: Loading spinner
+        case HashMap.lookup (Just category) state.frontpageFeeds of
+          Just feed -> setState _ { frontpageArticles = feed }
+          _ -> do
+            Aff.launchAff_ do
+              categoryFeed <- Lettera.getFrontpage HBL (Just $ show category)
+              liftEffect $
+                setState \s -> s { frontpageArticles = categoryFeed
```
Comment (kaol) [inline apps/mosaico/src/Mosaico.purs:155]: `setState _`

File: apps/mosaico/src/Mosaico.purs [hunk 4]
```diff
@@ -229,6 +267,15 @@ render setState state router =
         }
     _ -> mempty
   <> case state.route of
+       Routes.CategoryPage _category ->
+         mosaicoDefaultLayout $ state.frontpageComponent
+           { frontpageArticles: state.frontpageArticles
+           , onArticleClick: \article -> do
+               setState \s -> s { clickedArticle = Just article }
```
Comment (kaol) [inline apps/mosaico/src/Mosaico.purs:274]: `setState _`

File: packages/components/src/Lettera/Models.purs [hunk 5]
```diff
@@ -302,18 +305,46 @@ data CategoryType
...
   decodeJson json = do
     categoryTypeString <- decodeJson json
-    case String.toLower categoryTypeString of
+    case toLower categoryTypeString of
       "feed"    -> Right Feed
       "webview" -> Right Webview
       "link"    -> Right Link
       _         -> Left $ UnexpectedValue json
 
+newtype CategoryLabel = CategoryLabel String
```
Comment (kaol) [inline packages/components/src/Lettera/Models.purs:322]: Tags have a similar representation issue going on with them, they also need a "display" and a "URL" style of them.

**Review by kaol — APPROVED** [review]

File: apps/mosaico/src/Main.purs (hunk 1 above)
Comment (vapaj) [reply apps/mosaico/src/Main.purs:150]: Yes, this is a bit weird. I don't have a clue what kind of input would make this fail. I tried.

File: packages/components/src/Lettera/Models.purs (hunk 5 above)
Comment (vapaj) [reply packages/components/src/Lettera/Models.purs:322]: Right. Maybe we could have a more general type for these.

File: apps/mosaico/src/Mosaico.purs (hunk 2 above)
Comment (vapaj) [reply apps/mosaico/src/Mosaico.purs:126]: I'm using the `s` variable in this lambda

**Pull request merged**